rpc/chain/client.go

```diff
@@ -0,0 +1,691 @@
+package chain
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+
+	"github.com/afex/hystrix-go/hystrix"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/status-im/status-go/services/rpcstats"
+)
+
+type FeeHistory struct {
+	BaseFeePerGas []string `json:"baseFeePerGas"`
+}
+
+type ClientWithFallback struct {
+	ChainID  uint64
+	main     *ethclient.Client
+	fallback *ethclient.Client
+
+	mainRpc     *rpc.Client
+	fallbackRpc *rpc.Client
+
+	IsConnected bool
+}
+
+func NewSimpleClient(main *rpc.Client, chainID uint64) *ClientWithFallback {
+	hystrix.ConfigureCommand(fmt.Sprintf("ethClient_%d", chainID), hystrix.CommandConfig{
+		Timeout:               10000,
+		MaxConcurrentRequests: 100,
+		SleepWindow:           300000,
+		ErrorPercentThreshold: 25,
+	})
+
+	return &ClientWithFallback{
+		ChainID:     chainID,
+		main:        ethclient.NewClient(main),
+		fallback:    nil,
+		mainRpc:     main,
+		fallbackRpc: main,
+		IsConnected: true,
+	}
+}
+
+func NewClient(main, fallback *rpc.Client, chainID uint64) *ClientWithFallback {
+	hystrix.ConfigureCommand(fmt.Sprintf("ethClient_%d", chainID), hystrix.CommandConfig{
+		Timeout:               10000,
+		MaxConcurrentRequests: 100,
+		SleepWindow:           300000,
+		ErrorPercentThreshold: 25,
+	})
+
+	var fallbackEthClient *ethclient.Client
+	if fallback != nil {
+		fallbackEthClient = ethclient.NewClient(fallback)
+	}
+	return &ClientWithFallback{
+		ChainID:     chainID,
+		main:        ethclient.NewClient(main),
+		fallback:    fallbackEthClient,
+		mainRpc:     main,
+		fallbackRpc: fallback,
+		IsConnected: true,
+	}
+}
+
+func (c *ClientWithFallback) Close() {
+	c.main.Close()
+	c.fallback.Close()
```

Comment:
This is allowed to be `nil` with `NewClient` constructor. I also see a double close in `NewSimpleClient` use case.

Reply:
Once again, good catch :) i will check for nil